hare3/remote.go

```diff
@@ -0,0 +1,257 @@
+package hare3
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/signing"
+)
+
+type NodeService interface {
+	GetHareMessage(ctx context.Context, layer types.LayerID, round IterRound) ([]byte, error)
+	Beacon(ctx context.Context, epoch types.EpochID) (types.Beacon, error)
+	Publish(ctx context.Context, proto string, blob []byte) error
+}
+
+type RemoteHare struct {
+	config    Config
+	wallClock clockwork.Clock
+	nodeClock nodeClock
+	mu        sync.Mutex
+	beacons   map[types.EpochID]types.Beacon
+	signers   map[string]*signing.EdSigner
+	oracle    *legacyOracle
+	sessions  map[types.LayerID]*protocol
+	eg        errgroup.Group
+	ctx       context.Context
+	svc       NodeService
+
+	log *zap.Logger
+}
+
+func NewRemoteHare(config Config,
+	nodeClock nodeClock,
+	nodeService NodeService,
+	oracle oracle,
+	log *zap.Logger,
+) *RemoteHare {
+	return &RemoteHare{
+		config:    config,
+		nodeClock: nodeClock,
+		beacons:   make(map[types.EpochID]types.Beacon),
+		signers:   make(map[string]*signing.EdSigner),
+		oracle: &legacyOracle{
+			log:    zap.NewNop(),
+			oracle: oracle,
+			config: DefaultConfig(),
+		},
+
+		sessions:  make(map[types.LayerID]*protocol),
+		eg:        errgroup.Group{},
+		svc:       nodeService,
+		log:       log,
+		wallClock: clockwork.NewRealClock(),
+	}
+}
+
+func (h *RemoteHare) Register(sig *signing.EdSigner) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.log.Info("registered signing key", log.ZShortStringer("id", sig.NodeID()))
+	h.signers[string(sig.NodeID().Bytes())] = sig
+}
+
+func (h *RemoteHare) Start(ctx context.Context) {
+	current := h.nodeClock.CurrentLayer() + 1
+	enabled := max(current, h.config.EnableLayer, types.GetEffectiveGenesis()+1)
+	disabled := types.LayerID(math.MaxUint32)
+	h.log.Info("started",
+		zap.Inline(&h.config),
+		zap.Uint32("enabled", enabled.Uint32()),
+		zap.Uint32("disabled", disabled.Uint32()),
```

Comment:
`enabled` and `disabled` names are slightly confusing, maybe `enabledAtLayer` / `disabledAtLayer`?

Reply:
this was carried over from the main hare implementation, kept the same logging style for uniformity